# LUMPY pipeline on hoffman2
## Overview
LUMPY is based upon a general probabilistic representation of an SV breakpoint that allows any number of alignment signals to be integrated into a single discovery process [1]. It can integrate multiple types of signals from a sample (as illustrated in A) and integreat a single signal types from multiple samples (as illustrated in B). 
![Integrate three signals from a sample](https://github.com/hjzhou988/Genome-STRiP-pipeline/blob/master/LUMPY/LUMPY_A.png )
![Integrate a single signal types from multiple samples](https://github.com/hjzhou988/Genome-STRiP-pipeline/blob/master/LUMPY/LUMPY_B.png)
In scenario A, LUMPY has its own Read-Pair module and Split-Read module, which requires the "discordant read pair" bam file and "split read" bam file as inputs. The Generic module needs the output from other algorithms or known variants. The output from other algorithms or known variants need to be convert to "bedpe" format by the script "vcfToBedpe" or "cnvanator_to_bedpes.py" (data generated by CNVnator) on LUMPY github. 
In scenario B, multiple samples (typically a trio) are used as inputs for one LUMPY run.

In the pipeline described here, the majory part is directly from [Colin Farrell's work](https://github.com/NuttyLogic/SV_Pipeline), with some modifications and annotations. Install LUMPY from its github page https://github.com/arq5x/lumpy-sv. 

## Single sample SV calling by tradition LUMPY
Example script is in "lumpy_variants_HZ.sh" and "lumpy_submission_shell_HZ.sh". 
Do not use lumpyexpress, as it will not work with the downstream sorting and merging multiple VCF files.

## Sort and merge single-sample VCFs
Use the LUMPY provided l_sort.py and l_merge.py scripts to sort and merge all single-sample VCF together. 
l_sort.py needs all single-sample VCFs as input. 
l_merge.py needs the output from l_sort.py.
It consumes a lot of memory. The job submission script uses 33G memory for l_merge.py and a little less for l_sort.py.

## Regions to exclude
The LUMPY paper has its own regions to exclude, based on certain regions having abnormally high coverage (https://github.com/hall-lab/speedseq/blob/master/annotations/ceph18.b37.lumpy.exclude.2014-01-15.bed). 
The LUMPY github also recommends the exclusion region published by Heng Li (https://doi.org/10.1093/bioinformatics/btu356)
The exclusion regions can be used either in the LUMPY running step. Otherwise you can use "bedtools intersect" function to delete the sites that intersect with the exclusion regions. 

## Genotyping



[1] Layer, Ryan M., et al. "LUMPY: a probabilistic framework for structural variant discovery." Genome biology 15.6 (2014): R84.

